kythe/rust/indexer/src/providers.rs

```diff
@@ -0,0 +1,152 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::KytheError;
+use analysis_rust_proto::*;
+use positioned_io;
+use rc_zip::{prelude::*, Archive, EntryContents};
+use std::fs::File;
+use std::io::Read;
+
+/// A trait for retrieving files during indexing.
+pub trait FileProvider {
+  /// Checks whether a file exists and that the contents is available.
+  ///
+  /// This function takes a file hash and returns whether
+  /// the file exists and that the contents can be requested.
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a [KytheError::KzipFileError] will
+  /// be returned.
+  fn exists(&mut self, file_hash: &str) -> bool;
+
+  /// Retrieves the string contents of a file.
+  ///
+  /// This function takes a file hash and returns the string contents of the
+  /// file.
+  ///
+  /// # Errors
+  ///
+  /// If the file does not exist, a
+  /// [FileNotFoundError][KytheError::FileNotFoundError] will be returned.
+  /// If the file can't be read into a String, a
+  /// [FileReadError][KytheError::FileReadError] will be returned.
+  fn contents(&mut self, file_hash: &str) -> Result<String, KytheError>;
+}
+
+/// A [FileProvider] that backed by a .kzip file.
+pub struct KzipFileProvider {
+  file: File,
+  zip_archive: Archive,
+}
+
+impl KzipFileProvider {
+  /// Create a new instance using a valid [File] object of a kzip
+  /// file.KytheError
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a
+  /// [KzipFileError][KytheError::KzipFileError] will be returned.
+  pub fn new(file: File) -> Result<Self, KytheError> {
+    let zip_archive = match file.read_zip() {
+      Ok(zip_archive) => zip_archive,
+      Err(_e) => return Err(KytheError::KzipFileError),
+    };
+    Ok(Self { file, zip_archive })
+  }
+
+  /// Retrieve the Compilation Units from the kzip.
+  ///
+  /// This function will attempt to parse the proto files in the `pbunits`
+  /// folder into Compilation Units, returning a Vector of them.
+  ///
+  /// # Errors
+  ///
+  /// If a file cannot be parsed, a
+  /// [ProtobufParseError][KytheError::ProtobufParseError] will be returned.
+  pub fn get_compilation_units(&self) -> Result<Vec<CompilationUnit>, KytheError> {
+    let mut compilation_units = Vec::new();
+    for entry in self.zip_archive.entries() {
+      // Protobuf files are in the pbunits folder
+      if entry.name().contains("/pbunits/") {
+        // Read the file into a compilation unit protobuf
+        match entry.contents() {
```

Comment:
Have you run [rust-clippy](https://github.com/rust-lang/rust-clippy) on this project? If not, I recommend using it -- it would have suggested [`single_match`](https://rust-lang.github.io/rust-clippy/master/#single_match) here.

Reply:
Ah interesting! I didn't know there was a linter. I'll install it and include it in my workflow. Thank you!

Reply:
Having trouble setting up rust-clippy because of Bazel. I might have to write a custom Bazel rule to support linting, but that will have to be a TODO. For now I will fix that code manually.

Reply:
It might be worth asking on the internal mailing list about this. Clippy support feels like a common enough requirement that someone may have looked into it already.

Reply:
Working on it actively in https://github.com/bazelbuild/rules_rust/issues/304

Reply:
Clippy support is now in the Rust rules. The next PR will contain a Bazel target for it.